selvpcclient/httpclient.py

```diff
@@ -44,14 +55,148 @@ def request(self, method, url, **kwargs):
+        )
+
+    def patch(self, path: str, service: str, region: str, **kwargs) -> Resp:
+        return self.request(
+            'PATCH', self.make_url(service, region, path), **kwargs
+        )
+
+    def delete(self, path: str, service: str, region: str, **kwargs) -> Resp:
+        return self.request(
+            'DELETE', self.make_url(service, region, path), **kwargs
+        )
+
+    def make_url(self, service: str, region: str, path: str) -> str:
+        return f'{self.identity.get_url_by_service(service, region)}{path}'
+
+
+@dataclass
+class XAuthToken:
+    token: str
+    expires: dt = None
+
+
+class IdentityManager:
+    MIN_TOKEN_TTL = 180
+
+    keystone: KeystoneClient = None
+    x_auth_token: XAuthToken = None
+    account_name: str = None
+
+    def __init__(
+        self,
+        cloud_management: HTTPClient,
+        identity_url: str,
+        account_name: str
+    ):
+        self.cloud_management = cloud_management
+        self.identity_url = identity_url
+        self.account_name = account_name
+
+    def get_x_auth_token(self) -> str:
+        if not self.x_auth_token:
+            self.x_auth_token = self._issue_x_auth_token()
+            return self.x_auth_token.token
+
+        ttl = self.x_auth_token.expires.timestamp() - dt.now().timestamp()
+        if ttl < self.MIN_TOKEN_TTL:
+            self.x_auth_token = self._issue_x_auth_token()
+            return self.x_auth_token.token
+
+        return self.x_auth_token.token
+
+    def get_url_by_service(self, service: str, region: str) -> str:
+        x_auth_token = self.get_x_auth_token()
+        token_info = self.keystone.tokens.validate(token=x_auth_token)
+
+        return token_info.service_catalog.url_for(
+            service_type=service,
+            region_name=region,
+        )
+
+    def _issue_x_auth_token(self) -> XAuthToken:
+        resp = self.cloud_management.post(
+            path='/tokens',
+            json={'token': {'account_name': self._get_account_name()}},
+        )
+        x_auth_token = resp.json()['token']['id']
+
+        # Re-initialize keystone client because the token has been refreshed.
+        self._init_keystone_client(token=x_auth_token)
+
+        token_info = self.keystone.tokens.validate(token=x_auth_token)
+
+        return XAuthToken(token=x_auth_token, expires=token_info.expires)
+
+    def _get_account_name(self) -> str:
+        if self.account_name:
+            return self.account_name
+
+        try:
+            x_token = self.cloud_management.headers['X-Token']
+        except KeyError:
+            raise ClientException(message='X-Token not set.')
+
+        if not re.match(r"\w{25}_\d{1,7}", x_token):
+            raise ClientException(
+                message=(
+                    'Invalid X-Token format. '
+                    'Token must have the following pattern: \\w{25}_\\d{1,7}'
+                )
+            )
+
+        self.account_name = x_token.split('_')[1]
```

Comment:
Why you try to get account name if you already have environment variable

Reply:
For backwards compatibility.
Now account name is retrieved via /accounts. This code has been removed.